lib/Driver/Compilation.cpp

```diff
@@ -920,6 +922,119 @@ namespace driver {
       return false;
     }
 
+    // Selects the number of partitions based on the user-provided batch
+    // count and/or the number of parallel tasks we can run, subject to a
+    // fixed per-batch safety cap, to avoid overcommitting memory.
+    size_t pickNumberOfPartitions() {
+
+      // If the user asked for something, use that.
+      if (Comp.getBatchCount().hasValue())
+        return Comp.getBatchCount().getValue();
+
+      // This is a long comment to justify a simple calculation.
+      //
+      // Because there is a secondary "outer" build system potentially also
+      // scheduling multiple drivers in parallel on separate build targets
+      // -- while we, the driver, schedule our own subprocesses -- we might
+      // be creating up to $NCPU^2 worth of _memory pressure_.
+      //
+      // Oversubscribing CPU is typically no problem these days, but
+      // oversubscribing memory can lead to paging, which on modern systems
+      // is quite bad.
+      //
+      // In practice, $NCPU^2 processes doesn't _quite_ happen: as core
+      // count rises, it usually exceeds the number of large targets (which
+      // are the only thing we have to worry about): you might have (say) 2
+      // large modules * 2 architectures, but that's only an $NTARGET value
+      // of 4, which is much less than $NCPU if you're on a 24 or 36-way
```

Comment:
It's not exactly "number of large targets"; it's "number of large targets that don't depend on one another". I think it's worth defining that carefully.

Reply:
fixed